network/src/protocols/test.rs

```diff
@@ -0,0 +1,386 @@
+
+    // Ping protocol
+    let (ping_sender, ping_receiver) = channel(std::u8::MAX as usize);
+    let ping_interval = Duration::from_secs(5);
+    let ping_timeout = Duration::from_secs(10);
+
+    let ping_meta = MetaBuilder::default()
+        .id(PING_PROTOCOL_ID.into())
+        .service_handle(move || {
+            ProtocolHandle::Both(Box::new(PingHandler::new(
+                ping_interval,
+                ping_timeout,
+                ping_sender.clone(),
+            )))
+        })
+        .build();
+
+    // Discovery protocol
+    let (disc_sender, disc_receiver) = mpsc::unbounded();
+    let disc_meta = MetaBuilder::default()
+        .id(DISCOVERY_PROTOCOL_ID.into())
+        .service_handle(move || {
+            ProtocolHandle::Both(Box::new(DiscoveryProtocol::new(disc_sender.clone())))
+        })
+        .build();
+
+    // Identify protocol
+    let identify_callback = IdentifyCallback::new(Arc::clone(&network_state), name);
+    let identify_meta = MetaBuilder::default()
+        .id(IDENTIFY_PROTOCOL_ID.into())
+        .service_handle(move || {
+            ProtocolHandle::Both(Box::new(IdentifyProtocol::new(identify_callback.clone())))
+        })
+        .build();
+
+    // Feeler protocol
+    let feeler_meta = MetaBuilder::default()
+        .id(FEELER_PROTOCOL_ID.into())
+        .name(move |_| "/ckb/flr/".to_string())
+        .service_handle({
+            let network_state = Arc::clone(&network_state);
+            move || ProtocolHandle::Both(Box::new(Feeler::new(Arc::clone(&network_state))))
+        })
+        .build();
+
+    let service_builder = ServiceBuilder::default()
+        .insert_protocol(ping_meta)
+        .insert_protocol(disc_meta)
+        .insert_protocol(identify_meta)
+        .insert_protocol(feeler_meta);
+
+    let mut p2p_service = service_builder
+        .key_pair(network_state.local_private_key().clone())
+        .upnp(config.upnp)
+        .forever(true)
+        .build(EventHandler {
+            network_state: Arc::clone(&network_state),
+        });
+
+    let disc_service = DiscoveryService::new(
+        Arc::clone(&network_state),
+        disc_receiver,
+        config.discovery_local_address,
+    );
+
+    let ping_service = PingService::new(
+        Arc::clone(&network_state),
+        p2p_service.control().to_owned(),
+        ping_receiver,
+    );
+
+    let listen_addr = p2p_service
+        .listen("/ip4/0.0.0.0/tcp/0".parse().unwrap())
+        .unwrap();
+
+    let control = p2p_service.control().clone();
+
+    thread::spawn(move || {
+        let mut rt = tokio::runtime::Runtime::new().unwrap();
+        rt.spawn(disc_service);
+        rt.spawn(ping_service.for_each(|_| Ok(())));
+        rt.block_on(p2p_service.for_each(|_| Ok(())))
+    });
+
+    Node {
+        control,
+        listen_addr,
+        network_state,
+    }
+}
+
+fn assert_help(node: &Node, expect_num: usize, step_second: u64, max_round: u8) {
```

Comment:
https://github.com/nervosnetwork/ckb/blob/develop/test/src/utils.rs#L74

A similar and more general implementation.

Reply:
> https://github.com/nervosnetwork/ckb/blob/develop/test/src/utils.rs#L74
> 
> A similar and more general implementation.

but it can't reuse here, it on `test/` not util